share-coordinator/src/main/java/org/apache/kafka/coordinator/share/ShareCoordinatorShard.java

```diff
@@ -537,38 +539,195 @@ private static ShareGroupOffset merge(ShareGroupOffset soFar, ShareUpdateValue n
+                if (batch.firstOffset() >= cur.firstOffset() && batch.lastOffset() <= cur.lastOffset()) {
+                    // extra batch needs to be created
+                    if (batch.firstOffset() > cur.firstOffset()) {
+                        batchQueue.add(
+                            new PersisterStateBatch(
+                                cur.firstOffset(),
+                                batch.firstOffset() - 1,
+                                cur.deliveryState(),
+                                cur.deliveryCount()
+                            )
+                        );
+                    }
+
+                    // extra batch needs to be created
+                    if (batch.lastOffset() < cur.lastOffset()) {
+                        batchQueue.add(
+                            new PersisterStateBatch(
+                                batch.lastOffset() + 1,
+                                cur.lastOffset(),
+                                cur.deliveryState(),
+                                cur.deliveryCount()
+                            )
+                        );
+                    }
+                } else if (batch.firstOffset() < cur.firstOffset() && batch.lastOffset() < cur.lastOffset()) {
+                    // covers
+                    //  ______    
+                    //____ 
+                    batchQueue.add(new PersisterStateBatch(
+                        batch.lastOffset() + 1,
+                        cur.lastOffset(),
+                        cur.deliveryState(),
+                        cur.deliveryCount()
+                    ));
+                } else if (batch.firstOffset() > cur.firstOffset() && batch.lastOffset() > cur.lastOffset()) {
+                    // covers
+                    //  ______  
+                    //   _______    
+                    batchQueue.add(new PersisterStateBatch(
+                        cur.firstOffset(),
+                        batch.firstOffset() - 1,
+                        cur.deliveryState(),
+                        cur.deliveryCount()
+                    ));
+                }
+            }
+            batchQueue.add(batch);
+        }
+
+        // sort, prune and merge intervals
+        return mergeBatches(
+            pruneBatches(
+                getSortedList(batchQueue),
+                startOffset
+            )
+        );
+    }
+
+    private static List<PersisterStateBatch> getSortedList(Collection<PersisterStateBatch> collection) {
+        List<PersisterStateBatch> finalList = new ArrayList<>(collection);
+        finalList.sort(PersisterStateBatch::compareTo);
+        return finalList;
+    }
+
+    // assumes sorted input
+    private static List<PersisterStateBatch> mergeBatches(List<PersisterStateBatch> batches) {
+        if (batches.size() < 2) {
+            return batches;
+        }
+        Stack<PersisterStateBatch> stack = new Stack<>();
+        stack.add(batches.get(0));
```

Comment:
Can't this be `batches.remove(0)`? Otherwise, the first element will be added to the stack and also the initial candidate in the loop.